HOMEWORK 2

You will be working with JSON data and MongoDB in this homework. The overall task is to query the bike sharing app Citi Bike NYC server through their API ( https://gbfs.lyft.com/gbfs/2.3/bkn/en/station_status.json ) for station status, store the data first in json files, then load the data into a MongoDB database, and finally infer activities for the stations. Sample starter code relating to these subtasks was discussed in class.

If you are interested in learning more about the data or the app, you can find information at https://citibikenyc.com/system-data and https://citibikenyc.com, respectively. This is not needed for the assignment. 


You are required to query the server at 3 different times on the same day (preferably a weekday). Make sure that these times are spread apart by around 3-5 hours. 3 good times may be morning, afternoon, and evening. Set up the code beforehand and a reminder on your phones to run the code :-). 

You should query the data, save the results in 3 JSON files on your local drive, load the JSON files as JSON Objects, and then insert them into the database. Your json files should be named <i>citibikenyc-1.json</i>, <i>citibikenyc-2.json</i> , and <i>citibikenyc-3.json</i> . You need to load the data into MongoDB according to the instructions in this notebook.

Make sure that your code is organized, documented, and contains only what is being asked for in the question. 

This homework requires you to run your data collection code at three different times of a day (preferably a weekday), so make sure that you start the homework early. 

Rename this file as <i>YourfirstinitialLastname.ipynb</i> and insert your code in the appropriate parts. For example, John Doe would name his solution file as <i>JDoe.ipynb</i> . 


Your submission should consist of both the ipynb file and the 3 json data files. 

# Part 1 (20 points)

## Get Dataset 1

Query the Citi Bike server through the API https://gbfs.lyft.com/gbfs/2.3/bkn/en/station_status.json and load the stations information in the result into a JSON Array. Note that the code discussed in the class may have shown an earlier version of the API. You must use the API given in this notebook for the assignment. 

The JSON Array will have JSON objects containing information for each station as shown below. The id of the station is stored in a <b>string</b> field named <b>station_id</b>.  

{'num_ebikes_available': 2,
 'num_bikes_disabled': 1,
 'num_docks_disabled': 0,
 'is_installed': 1,
 'num_scooters_available': 0,
 'last_reported': 1708640357,
 'is_returning': 1,
 'station_id': 'fbaffdbd-dab5-4127-9bd5-36bf58093ea8',
 'vehicle_types_available': [{'vehicle_type_id': '1', 'count': 13},
  {'vehicle_type_id': '2', 'count': 2}],
 'is_renting': 1,
 'num_scooters_unavailable': 0,
 'num_docks_available': 8,
 'num_bikes_available': 15}
 
 Display the entries for the first and 10th stations in the JSON array.

In [17]:
# do any necessary imports
import json
import requests
import pandas as pd

In [19]:
#get data for citibike nyc from API

#this data was pulled on Sunday as I ran the query multiple times, 
#but the analysis done below was from saved data which was pulled on weekday 3 times in span of 5 hours

response1 = requests.get("https://gbfs.lyft.com/gbfs/2.3/bkn/en/station_status.json")

In [146]:
# load data into json array
stations1 = json.loads(response1.content.decode('utf-8'))['data']['stations']

In [147]:
# print data for first station
stations1[0]

{'is_returning': 1,
 'num_docks_disabled': 0,
 'station_id': '66db9e99-0aca-11e7-82f6-3863bb44ef7c',
 'is_installed': 1,
 'num_scooters_available': 0,
 'num_docks_available': 7,
 'num_bikes_available': 11,
 'vehicle_types_available': [{'count': 4, 'vehicle_type_id': '1'},
  {'count': 7, 'vehicle_type_id': '2'}],
 'num_scooters_unavailable': 0,
 'num_ebikes_available': 7,
 'num_bikes_disabled': 1,
 'last_reported': 1709483020,
 'is_renting': 1}

In [148]:
# print data for tenth station
stations1[9]

{'is_returning': 1,
 'num_docks_disabled': 0,
 'station_id': 'ca5cc034-a95d-446e-8314-6691589886b1',
 'is_installed': 1,
 'num_scooters_available': 0,
 'num_docks_available': 7,
 'num_bikes_available': 11,
 'vehicle_types_available': [{'count': 11, 'vehicle_type_id': '1'},
  {'count': 0, 'vehicle_type_id': '2'}],
 'num_scooters_unavailable': 0,
 'num_ebikes_available': 0,
 'num_bikes_disabled': 1,
 'last_reported': 1709483020,
 'is_renting': 1}

## Load Dataset 1 into Pandas DataFrame
Load the JSON Array from the previous section into a Pandas DataFrame and display the first 10 rows. 

In [149]:
stations_df = pd.DataFrame(stations1)

In [150]:
stations_df.head(10)

is_returning  num_docks_disabled                            station_id  \
0             1                   0  66db9e99-0aca-11e7-82f6-3863bb44ef7c   
1             1                   0  d73e26de-8534-4b46-a321-3cae295ef9b2   
2             1                   0                   1860188747953293684   
3             1                   0  66dde76a-0aca-11e7-82f6-3863bb44ef7c   
4             1                   0  a45a712c-1e3d-4fee-bef8-adafafd80670   
5             1                   0  66dd4ccc-0aca-11e7-82f6-3863bb44ef7c   
6             1                   0  00967b8f-1a4d-4131-a65a-17fa8ca89e28   
7             1                   0  8a89684b-40a5-49fe-b75d-fdf32ffa8102   
8             1                   0  d979131e-0b9f-4aa2-88c5-46d4ff94abc8   
9             1                   0  ca5cc034-a95d-446e-8314-6691589886b1   

   is_installed  num_scooters_available  num_docks_available  \
0             1                     0.0                    7   
1             1                     0.0                   13   
2             1                     0.0                    8   
3             1                     0.0                   16   
4             1                     0.0                    0   
5             1                     0.0                   23   
6             1                     0.0                    6   
7             1                     0.0                   11   
8             1                     0.0                    8   
9             1                     0.0                    7   

   num_bikes_available                            vehicle_types_available  \
0                   11  [{'count': 4, 'vehicle_type_id': '1'}, {'count...   
1                   13  [{'count': 0, 'vehicle_type_id': '1'}, {'count...   
2                   10  [{'count': 3, 'vehicle_type_id': '1'}, {'count...   
3                    9  [{'count': 4, 'vehicle_type_id': '1'}, {'count...   
4                   20  [{'count': 19, 'vehicle_type_id': '1'}, {'coun...   
5                   27  [{'count': 0, 'vehicle_type_id': '1'}, {'count...   
6                   14  [{'count': 13, 'vehicle_type_id': '1'}, {'coun...   
7                    8  [{'count': 5, 'vehicle_type_id': '1'}, {'count...   
8                   10  [{'count': 6, 'vehicle_type_id': '1'}, {'count...   
9                   11  [{'count': 11, 'vehicle_type_id': '1'}, {'coun...   

   num_scooters_unavailable  num_ebikes_available  num_bikes_disabled  \
0                       0.0                     7                   1   
1                       0.0                    13                   4   
2                       0.0                     7                   1   
3                       0.0                     5                   0   
4                       0.0                     1                   0   
5                       0.0                    27                   1   
6                       0.0                     1                   0   
7                       0.0                     3                   1   
8                       0.0                     4                   1   
9                       0.0                     0                   1   

   last_reported  is_renting  
0     1709483020           1  
1     1709483020           1  
2     1709483020           1  
3     1709483020           1  
4     1709483020           1  
5     1709483011           1  
6     1709483011           1  
7     1709483012           1  
8     1709483012           1  
9     1709483020           1

## Dataset Details

Run the info() and describe() methods of DataFrames and answer the questions below from the results of these methods. Markdown cells have been inserted for you for answering these questions.

(1) How many stations are there in total?

(2) How many fields are available for each station?

(3) What is the maximum value for <i>num_bikes_available</i> across all stations?

In [178]:
stations_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2209 entries, 0 to 2208
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   is_returning              2209 non-null   int64  
 1   num_docks_disabled        2209 non-null   int64  
 2   station_id                2209 non-null   object 
 3   is_installed              2209 non-null   int64  
 4   num_scooters_available    2146 non-null   float64
 5   num_docks_available       2209 non-null   int64  
 6   num_bikes_available       2209 non-null   int64  
 7   vehicle_types_available   2209 non-null   object 
 8   num_scooters_unavailable  2146 non-null   float64
 9   num_ebikes_available      2209 non-null   int64  
 10  num_bikes_disabled        2209 non-null   int64  
 11  last_reported             2209 non-null   int64  
 12  is_renting                2209 non-null   int64  
dtypes: float64(2), int64(9), object(2)
memory usage: 224.5+ KB


In [179]:
stations_df.describe()

is_returning  num_docks_disabled  is_installed  num_scooters_available  \
count   2209.000000         2209.000000   2209.000000                  2146.0   
mean       0.991399            0.228610      0.994568                     0.0   
std        0.092364            1.456736      0.073520                     0.0   
min        0.000000            0.000000      0.000000                     0.0   
25%        1.000000            0.000000      1.000000                     0.0   
50%        1.000000            0.000000      1.000000                     0.0   
75%        1.000000            0.000000      1.000000                     0.0   
max        1.000000           29.000000      1.000000                     0.0   

       num_docks_available  num_bikes_available  num_scooters_unavailable  \
count          2209.000000          2209.000000                    2146.0   
mean             13.795835            15.255319                       0.0   
std              11.981699            14.699681                       0.0   
min               0.000000             0.000000                       0.0   
25%               5.000000             5.000000                       0.0   
50%              12.000000            12.000000                       0.0   
75%              19.000000            20.000000                       0.0   
max              97.000000           118.000000                       0.0   

       num_ebikes_available  num_bikes_disabled  last_reported   is_renting  
count           2209.000000         2209.000000   2.209000e+03  2209.000000  
mean               4.643730            1.411951   1.704834e+09     0.991399  
std                7.656774            1.700267   8.898719e+07     0.092364  
min                0.000000            0.000000   8.640000e+04     0.000000  
25%                0.000000            0.000000   1.709483e+09     1.000000  
50%                2.000000            1.000000   1.709483e+09     1.000000  
75%                6.000000            2.000000   1.709483e+09     1.000000  
max               76.000000           13.000000   1.709483e+09     1.000000

(1) How many stations are there in total?

Looking at the non null values for station ids provided in info, we can see that there are total 2209 stations 
We can also check the same by running len(station)

(2) How many fields are available for each station?

There are total 13 fields available for each station, however "num_scooters_unavailable" & "num_scooters_available" are either null or not available in some of the stations

(3) What is the maximum value for num_bikes_available across all stations?

Across all stations, 118 is the maximum value for num_bikes_available

## Save the dataset

Save the JSON data that you collected in a file named citibikenyc-1.json on your local drive. 

In [9]:
# Dataset 3 taken at ~ 9:00 AM. 27 Feb 2024
with open('citibikenyc-1.json', 'w') as f:
        json.dump(stations1, f)

# Part 2  (20 points)

## Get Dataset 2

(1) Query the citibikenyc server again around 3-5 hours after the previous query on the same day. 

(2) Load the the station information as JSON data (array).

(3) Print out the details for the first and 10th stations.

(4) Save the JSON data in a file named citibikenyc-2.json on your local drive.

In [11]:
# Dataset 2 taken at ~ 2:00 PM. 27 Feb 2024
response2 = requests.get("https://gbfs.lyft.com/gbfs/2.3/bkn/en/station_status.json")

In [14]:
stations2 = json.loads(response.content.decode('utf-8'))['data']['stations']

In [16]:
with open('citibikenyc-2.json', 'w') as f:
        json.dump(stations2, f)

## Get Dataset 3

(1) Query the citibikenyc server again around 3-5 hours after the previous query on the same day.

(2) Load the the station information as JSON data.

(3) Print out the details for the first and 10th stations.

(4) Save the JSON data in a file named citibikenyc-3.json on your local drive.

In [20]:
# Dataset 3 taken at ~ 7:00 PM. 27 Feb 2024
response3 = requests.get("https://gbfs.lyft.com/gbfs/2.3/bkn/en/station_status.json")

In [23]:
stations3 = json.loads(response3.content.decode('utf-8'))['data']['stations']

In [24]:
with open('citibikenyc-3.json', 'w') as f:
        json.dump(stations3, f)

# Part 3  (35 points)

## Open a connection to a MongoDB Collection

Install pymongo on your laptop/environment if you need to (you need to do this only once). 

Import the MongoClient from the pymongo package.

Set up environment variables and get a connection to a new collection (you can name it citibikenyc). 

In [ ]:
pip install pymongo

In [4]:
from pymongo import MongoClient
import pymongo

In [40]:
client = pymongo.MongoClient("mongodb+srv://dbms:dbms2024@cluster0.5byfqtj.mongodb.net/")

In [166]:
db = client["citibike"]
collection1 = db["collection1"]

## Load Dataset 1 and insert in MongoDB Collection

Load the citibikenyc-1.json file into memory and insert the data into a MongoDB collection with the following modifications:

The fields <b>num_bikes_available</b> and <b>last_reported</b> should both be arrays of their singleton values in the collection. For example, a sample station would look like the following in the collection:

{'_id': ObjectId('65d7ca64523b0897863fc6fa'), 'num_ebikes_available': 21, 'num_bikes_disabled': 6, 'num_docks_disabled': 0, 'is_installed': 1, 'num_scooters_available': 0, <b>'last_reported': [1708640402]</b>, 'is_returning': 1, 'station_id': '1827839088308194240', 'vehicle_types_available': [{'vehicle_type_id': '1', 'count': 92}, {'vehicle_type_id': '2', 'count': 21}], 'is_renting': 1, 'num_scooters_unavailable': 0, 'num_docks_available': 1,  <b>'num_bikes_available': [113] </b>}

(The last_reported field is in Epoch time, a common way of expressing timestamps in datas. Try to understand what this time means through an internet search.)

In [167]:
with open('citibikenyc-1.json') as file:
    data1 = json.load(file)

In [168]:
for stations in data1:
    stations['num_bikes_available'] = [stations['num_bikes_available']]
    stations['last_reported'] = [stations['last_reported']]

In [ ]:
collection1.insert_many(data1)

In [170]:
sample_documents = collection1.find().limit(3) 

for document in sample_documents:
    print(document)

{'_id': ObjectId('65e4f52c30ebec28f9400a13'), 'num_docks_available': 0, 'num_docks_disabled': 0, 'last_reported': [86400], 'num_bikes_disabled': 0, 'vehicle_types_available': [{'count': 0, 'vehicle_type_id': '1'}, {'count': 0, 'vehicle_type_id': '2'}], 'is_installed': 0, 'num_ebikes_available': 0, 'station_id': '66dc2995-0aca-11e7-82f6-3863bb44ef7c', 'num_bikes_available': [0], 'is_returning': 0, 'is_renting': 0}
{'_id': ObjectId('65e4f52c30ebec28f9400a14'), 'num_docks_available': 0, 'num_docks_disabled': 0, 'last_reported': [86400], 'num_bikes_disabled': 0, 'vehicle_types_available': [{'count': 0, 'vehicle_type_id': '1'}, {'count': 0, 'vehicle_type_id': '2'}], 'is_installed': 0, 'num_ebikes_available': 0, 'station_id': '06439006-11b6-44f0-8545-c9d39035f32a', 'num_bikes_available': [0], 'is_returning': 0, 'is_renting': 0}
{'_id': ObjectId('65e4f52c30ebec28f9400a15'), 'num_docks_available': 19, 'num_docks_disabled': 0, 'last_reported': [1709042158], 'num_bikes_disabled': 2, 'vehicle_typ

## Query the MongoDB collection 

Query the MongoDB collection to return all records where the <b>first element of the array field num_bikes_available</b> is greater than 100. Print the results. If you don't get any results for 100, try a smaller threshold until you get a few (around 1-3) results. 

A record should look like the example shown in Part 3.2.

In [162]:
result = collection1.find({"num_bikes_available": {"$gt": 100}})

for station in result:
    print(station)

{'_id': ObjectId('65e4f45830ebec28f94004c4'), 'num_docks_available': 0, 'num_docks_disabled': 0, 'last_reported': [1709042265], 'num_bikes_disabled': 1, 'vehicle_types_available': [{'count': 98, 'vehicle_type_id': '1'}, {'count': 20, 'vehicle_type_id': '2'}], 'is_installed': 1, 'num_scooters_available': 0, 'num_ebikes_available': 20, 'num_scooters_unavailable': 0, 'station_id': '1827839088308194240', 'num_bikes_available': [118], 'is_returning': 1, 'is_renting': 1}
{'_id': ObjectId('65e4f45830ebec28f940065b'), 'num_docks_available': 16, 'num_docks_disabled': 0, 'last_reported': [1709042261], 'num_bikes_disabled': 1, 'vehicle_types_available': [{'count': 93, 'vehicle_type_id': '1'}, {'count': 10, 'vehicle_type_id': '2'}], 'is_installed': 1, 'num_scooters_available': 0, 'num_ebikes_available': 10, 'num_scooters_unavailable': 0, 'station_id': '66dbe848-0aca-11e7-82f6-3863bb44ef7c', 'num_bikes_available': [104], 'is_returning': 1, 'is_renting': 1}


## Load Dataset 2 and Update Documents in MongoDB

Merge the collected data for the fields <b>num_bikes_available</b> and <b>last_reported</b> for each station in Dataset 2 from the file citibikenyc-2.json into the collection by appending to the appropriate arrays in the collection. You only need to merge these two fields. You can ignore other fields. 

It is possible, but not likely, that a station may be added or deleted between different queries. You should guard against a new station by dropping it i.e. not add the data to the collection. You don't need to guard against deletion in this assignment. 

IMPORTANT: Define a function to do this merge so that you can reuse it for Dataset 3 in the next section. 

Note that this operation may take a while to complete. 

In [180]:
def merge_data(file_name, collection1):
    with open(file_name, 'r') as file:
        data2 = json.load(file)

    for station in data2:
        query = {'station_id': station['station_id']}
        existing_document = collection1.find_one(query)

        if existing_document:
            # Merge data from collection1 first
            collection1.update_one(
                query,
                {
                    '$set': {
                        'num_bikes_available': existing_document['num_bikes_available'], 
                        'last_reported': existing_document['last_reported']
                    }
                }
            )
            collection1.update_one(
                query,
                {
                    '$push': {
                        'num_bikes_available': station['num_bikes_available'], 
                        'last_reported': station['last_reported']
                    }
                }
            )
        else:
            print(f"Station {station['station_id']} not found in collection1")
            # Handle the case when the station is not found in collection1
            continue
            
merge_data("citibikenyc-2.json", collection1)

In [172]:
sample_documents = collection1.find().limit(5)

for document in sample_documents:
    print(document)

{'_id': ObjectId('65e4f52c30ebec28f9400a13'), 'num_docks_available': 0, 'num_docks_disabled': 0, 'last_reported': [86400, 86400], 'num_bikes_disabled': 0, 'vehicle_types_available': [{'count': 0, 'vehicle_type_id': '1'}, {'count': 0, 'vehicle_type_id': '2'}], 'is_installed': 0, 'num_ebikes_available': 0, 'station_id': '66dc2995-0aca-11e7-82f6-3863bb44ef7c', 'num_bikes_available': [0, 0], 'is_returning': 0, 'is_renting': 0}
{'_id': ObjectId('65e4f52c30ebec28f9400a14'), 'num_docks_available': 0, 'num_docks_disabled': 0, 'last_reported': [86400, 86400], 'num_bikes_disabled': 0, 'vehicle_types_available': [{'count': 0, 'vehicle_type_id': '1'}, {'count': 0, 'vehicle_type_id': '2'}], 'is_installed': 0, 'num_ebikes_available': 0, 'station_id': '06439006-11b6-44f0-8545-c9d39035f32a', 'num_bikes_available': [0, 0], 'is_returning': 0, 'is_renting': 0}
{'_id': ObjectId('65e4f52c30ebec28f9400a15'), 'num_docks_available': 19, 'num_docks_disabled': 0, 'last_reported': [1709042158, 1709060286], 'num_

## Query the MongoDB collection 

Query the collection to return all records where the first element of the array field num_bikes_available is greater than 100. Print the results. If you don't get any results for 100, try a smaller threshold until you get a few (around 1-3) results. 

The data for a station will look like the following after the merge in the collection:

{'_id': ObjectId('65d7ca64523b0897863fc6fa'), 'num_ebikes_available': 21, 'num_bikes_disabled': 6, 'num_docks_disabled': 0, 'is_installed': 1, 'num_scooters_available': 0, 'last_reported': <b>[1708640402, 1708640764]</b>, 'is_returning': 1, 'station_id': '1827839088308194240', 'vehicle_types_available': [{'vehicle_type_id': '1', 'count': 92}, {'vehicle_type_id': '2', 'count': 21}], 'is_renting': 1, 'num_scooters_unavailable': 0, 'num_docks_available': 1, 'num_bikes_available': <b>[113, 16]</b>}

In [173]:
result = collection1.find({"num_bikes_available.0": {"$gt": 100}})

for station in result:
    print(station)

{'_id': ObjectId('65e4f52c30ebec28f9400d65'), 'num_docks_available': 0, 'num_docks_disabled': 0, 'last_reported': [1709042265, 1709060254], 'num_bikes_disabled': 1, 'vehicle_types_available': [{'count': 98, 'vehicle_type_id': '1'}, {'count': 20, 'vehicle_type_id': '2'}], 'is_installed': 1, 'num_scooters_available': 0, 'num_ebikes_available': 20, 'num_scooters_unavailable': 0, 'station_id': '1827839088308194240', 'num_bikes_available': [118, 115], 'is_returning': 1, 'is_renting': 1}
{'_id': ObjectId('65e4f52c30ebec28f9400efc'), 'num_docks_available': 16, 'num_docks_disabled': 0, 'last_reported': [1709042261, 1709060278], 'num_bikes_disabled': 1, 'vehicle_types_available': [{'count': 93, 'vehicle_type_id': '1'}, {'count': 10, 'vehicle_type_id': '2'}], 'is_installed': 1, 'num_scooters_available': 0, 'num_ebikes_available': 10, 'num_scooters_unavailable': 0, 'station_id': '66dbe848-0aca-11e7-82f6-3863bb44ef7c', 'num_bikes_available': [104, 115], 'is_returning': 1, 'is_renting': 1}


## Load Dataset 3 and update documents in MongoDB

Merge the collected data for the fields <b>num_bikes_available</b> and <b>last_reported</b> for each station in Dataset 3 from the file citibikenyc-3.json into the collection by appending to the appropriate arrays in the MongoDB collection. You only need to merge these two fields. You can ignore other fields. 

You should use the function you defined in the previous section to do this. 

Note that this operation may take a while to complete. 

Note: It is possible, but not likely, that a station may be added or deleted between different queries. You may get some errors if this happens. You can drop a newly added station i.e. not add the data to the collection. You don't need to guard against deletion in this assignment. 

In [174]:
merge_data("citibikenyc-3.json", collection1)

In [175]:
sample_documents = collection1.find().limit(3)

for document in sample_documents:
    print(document)

{'_id': ObjectId('65e4f52c30ebec28f9400a13'), 'num_docks_available': 0, 'num_docks_disabled': 0, 'last_reported': [86400, 86400, 86400], 'num_bikes_disabled': 0, 'vehicle_types_available': [{'count': 0, 'vehicle_type_id': '1'}, {'count': 0, 'vehicle_type_id': '2'}], 'is_installed': 0, 'num_ebikes_available': 0, 'station_id': '66dc2995-0aca-11e7-82f6-3863bb44ef7c', 'num_bikes_available': [0, 0, 0], 'is_returning': 0, 'is_renting': 0}
{'_id': ObjectId('65e4f52c30ebec28f9400a14'), 'num_docks_available': 0, 'num_docks_disabled': 0, 'last_reported': [86400, 86400, 86400], 'num_bikes_disabled': 0, 'vehicle_types_available': [{'count': 0, 'vehicle_type_id': '1'}, {'count': 0, 'vehicle_type_id': '2'}], 'is_installed': 0, 'num_ebikes_available': 0, 'station_id': '06439006-11b6-44f0-8545-c9d39035f32a', 'num_bikes_available': [0, 0, 0], 'is_returning': 0, 'is_renting': 0}
{'_id': ObjectId('65e4f52c30ebec28f9400a15'), 'num_docks_available': 19, 'num_docks_disabled': 0, 'last_reported': [1709042158

## Query the MongoDB collection 

Query the collection to return all records where the first element of the array field num_bikes_available is greater than 100. Print the results. If you don't get any results for 100, try a smaller threshold until you get a few (around 1-3) results. 

The data for a station will look like the following after the merge in the collection:

{'_id': ObjectId('65d7ca64523b0897863fc6fa'), 'num_ebikes_available': 21, 'num_bikes_disabled': 6, 'num_docks_disabled': 0, 'is_installed': 1, 'num_scooters_available': 0, 'last_reported': <b>[1708640402, 1708640764, 1708640764]</b>, 'is_returning': 1, 'station_id': '1827839088308194240', 'vehicle_types_available': [{'vehicle_type_id': '1', 'count': 92}, {'vehicle_type_id': '2', 'count': 21}], 'is_renting': 1, 'num_scooters_unavailable': 0, 'num_docks_available': 1, 'num_bikes_available': <b>[113, 16, 20]</b>}



In [176]:
result = collection1.find({"num_bikes_available.0": {"$gt": 100}})

for station in result:
    print(station)

{'_id': ObjectId('65e4f52c30ebec28f9400d65'), 'num_docks_available': 0, 'num_docks_disabled': 0, 'last_reported': [1709042265, 1709060254, 1709078298], 'num_bikes_disabled': 1, 'vehicle_types_available': [{'count': 98, 'vehicle_type_id': '1'}, {'count': 20, 'vehicle_type_id': '2'}], 'is_installed': 1, 'num_scooters_available': 0, 'num_ebikes_available': 20, 'num_scooters_unavailable': 0, 'station_id': '1827839088308194240', 'num_bikes_available': [118, 115, 110], 'is_returning': 1, 'is_renting': 1}
{'_id': ObjectId('65e4f52c30ebec28f9400efc'), 'num_docks_available': 16, 'num_docks_disabled': 0, 'last_reported': [1709042261, 1709060278, 1709078269], 'num_bikes_disabled': 1, 'vehicle_types_available': [{'count': 93, 'vehicle_type_id': '1'}, {'count': 10, 'vehicle_type_id': '2'}], 'is_installed': 1, 'num_scooters_available': 0, 'num_ebikes_available': 10, 'num_scooters_unavailable': 0, 'station_id': '66dbe848-0aca-11e7-82f6-3863bb44ef7c', 'num_bikes_available': [104, 115, 107], 'is_return

# Part V (25 points)¶

Your data contains two time intervals: the first time interval is between the times that Dataset 1 and Dataset 2 were collected and the second time interval is between the times that Dataset 2 and Dataset 3 were collected. Activity in an interval is defined as the absolute difference between the entries in the num_bikes_available array for the end points of the interval. E.g. the activities in the two time intervals shown for the sample in Section 3.7 would be 97 and 4, respectively. 

Query the MongoDB collection to compute the number of stations with high activity (>=20) and low activity (<=5) in each of the two time intervals. Print your results. They may look like the following (the actual numbers may differ):

High Activity Stations [10, 29]

Low Activity Stations [205, 308]

There will be 4 counts in total. A station with activity 25, 3 in the 2 time intervals will contribute towards the count for High Activity Stations in the first interval and contribute towards the count for the Low Activity Stations in the second interval. Note that general form of the matrix for the result is as follows:

High Activity Stations [x1, x2]

Low Activity Stations [y1, y2]

x1 and y1 are the counts for the first interval and x2 and y2 are the counts for the second interval. 

The station with activity: (25,3) will contribute to the counts in x1 and y2. 

 If there were just two stations with activities (25,3) and (30, 25), the results would be

          High Activity Stations [2, 1]
          Low Activity Stations [0, 1]

In [177]:
high_activity_counts = [0, 0]
low_activity_counts = [0, 0]

for stations in collection1.find():
    num_bikes_available = stations.get('num_bikes_available', [])

    if len(num_bikes_available) >= 3:
        difference1 = abs(num_bikes_available[1] - num_bikes_available[0])
        difference2 = abs(num_bikes_available[2] - num_bikes_available[1])

        # For the first time interval
        if difference1 >= 20:
            high_activity_counts[0] += 1
        elif difference1 <= 5:
            low_activity_counts[0] += 1

        # For the second time interval
        if difference2 >= 20:
            high_activity_counts[1] += 1
        elif difference2 <= 5:
            low_activity_counts[1] += 1

print("High Activity Stations", high_activity_counts)
print("Low Activity Stations", low_activity_counts)

High Activity Stations [42, 98]
Low Activity Stations [1873, 1699]


In [154]:
client.close()